### PR TITLE
Add details to the task struct

### PR DESCRIPTION
**Pull request by Diogenesoftoronto**
- `JobsListRequest` struct now comes with a detailed field to get more verbose output from the archivematica API.
- Breaking changes with the current API.
- Removes `TaskDetailed` and moves the fields in `TaskDetailed` to the `Task` Struct.
- Test for `Job` changed for detailed output.
- Types for `Task` struct changed.
- New custom unmarshall function for encapsulated type `Duration` added.

Refs artefactual-sdps/enduro#832

### CODE REVIEW

File: v2_jobs.go [hunk 1]
```diff
@@ -80,12 +83,17 @@ func (s *JobStatus) UnmarshalJSON(b []byte) error {
 }
 
 type Task struct {
-	ID       string `json:"uuid"`
-	ExitCode uint8  `json:"exit_code"`
+	ID          string `json:"uuid"`
+	ExitCode    uint8  `json:"exit_code"`
+	Filename    string `json:"file_name"`
+	CreatedAt   string `json:"time_created"`
```
Comment (djjuhasz) [inline v2_jobs.go]: I think the date/time values should be represented as a `time.Time` type. I believe `time.Time` supports automatic Unmarshalling of JSON RFC3339 dates: https://eli.thegreenplace.net/2020/unmarshaling-time-values-from-json/

**Review by djjuhasz — CHANGES_REQUESTED** [review]
@Diogenesoftoronto just one change request.

File: v2_jobs.go (hunk 1 above)
Comment (sevein) [reply v2_jobs.go]: You're right, David! We should reuse the existing struct.

Comment (djjuhasz) [reply v2_jobs.go]: Huh, if we already have `amclient.TaskDetailed` can't we just use that in jobs struct?  It seems redundant to have the same struct defined in two places.

File: v2_jobs.go [hunk 2]
```diff
@@ -80,12 +83,17 @@ func (s *JobStatus) UnmarshalJSON(b []byte) error {
...
+	StartedAt   string `json:"time_started"`
+	CompletedAt string `json:"time_ended"`
+	Duration    string `json:"duration"`
 }
 
 func (s *JobsServiceOp) List(ctx context.Context, ID string, r *JobsListRequest) ([]Job, *Response, error) {
-	path := fmt.Sprintf("%s/%s", jobsBasePath, ID)
+	path := fmt.Sprintf("%s/%s%s", jobsBasePath, ID, jobsDetailsQuery)
```
Comment (sevein) [inline v2_jobs.go]: This could or should be optional, e.g. a boolean attribute that is part of the request struct.

Use url.Values to encode query parameters. Here's a helper that could be used as a reference: https://github.com/digitalocean/godo/blob/c154e94a159362efd5f5b718cb49df029937ebd1/godo.go#L203.

File: v2_jobs_test.go [hunk 3]
```diff
@@ -23,8 +23,13 @@ func TestJobs_List(t *testing.T) {
...
 			{
-				"uuid": "491aebbd-457b-4a6e-adf6-87a3a9ee951a",
-				"exit_code": 1
+				"uuid": "491aebbd-457b-4a6e-adf6-87a3a9ee951a", 
+				"exit_code": 1,
+				"file_name": "Images-94ade01c-49ce-49e0-9cc3-805575c676d0",
+				"time_created": "2024-01-18T01:27:49",
+				"time_started": "2024-01-18T01:27:49",
+				"time_ended": "2024-01-18T01:27:49",
+				"duration": "< 1"
```
Comment (sevein) [inline v2_jobs_test.go:35]: We should not make up this payload, use an example returned by Archivematica. I don't think we should see a string here, should be an integer.

**Review by sevein — CHANGES_REQUESTED** [review]
I haven't tested but gave you some ideas.
From my phone, still sick! 🤒

File: v2_jobs_test.go (hunk 3 above)
Comment (Diogenesoftoronto) [reply v2_jobs_test.go:35]: This is copy pasted from archivematica

File: v2_jobs.go (hunk 1 above)
Comment (Diogenesoftoronto) [reply v2_jobs.go]: Honestly didn't know there was taskdetailed

File: v2_jobs.go (hunk 2 above)
Comment (Diogenesoftoronto) [reply v2_jobs.go]: I'll check this out.

File: v2_jobs.go (hunk 1 above)
Comment (Diogenesoftoronto) [reply v2_jobs.go]: > I think the date/time values should be represented as a `time.Time` type. I believe `time.Time` supports automatic Unmarshalling of JSON RFC3339 dates: https://eli.thegreenplace.net/2020/unmarshaling-time-values-from-json/

Thought of this. just thinking about time.Time to sql.Null which is what I have to convert this to in preservation task.

Comment (djjuhasz) [reply v2_jobs.go]: In `TaskDateTime` you could copy the implementation of [time.UnmarshalJSON()](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/time/time.go;l=1356) and return a zero time if the JSON value is "null" (or an empty string?):

```
func (t *TaskDateTime) UnmarshalJSON(data []byte) error {
        if string(data) == "null" {
             return nil
        }
	s := string(data)
	...
}
```

Then in the preservation task you could convert the `amclient.TaskDateTime` value to a `sql.NullTime` value.

Comment (djjuhasz) [reply v2_jobs.go]: Huh, from https://github.com/golang/go/issues/47353 it looks like the `time.UnmarshalJSON()` implementation actually has some issues with escaped Unicode.  Maybe it would be better to implement `encoding.TextUnmarshaler` instead?

File: v2_jobs_test.go (hunk 3 above)
Comment (sevein) [reply v2_jobs_test.go:35]: The API returns "< 1"? Wow. Then I guess we'll need a custom decoder that knows how to read that.

Comment (Diogenesoftoronto) [reply v2_jobs_test.go:35]: yeah i thought it was weird when i saw it but yeah it actually returns that :laughing:

File: v2_jobs.go (hunk 1 above)
Comment (Diogenesoftoronto) [reply v2_jobs.go]: > Huh, if we already have `amclient.TaskDetailed` can't we just use that in jobs struct? It seems redundant to have the same struct defined in two places.

	Duration    uint32       `json:"duration"`
task detailed has a duration as a uint32 which it is not. we can create a custom parse function to fix this or create a custom type for it or something call _archivematicaTime_ or something haha.

File: v2_task.go [hunk 4]
```diff
@@ -22,6 +24,36 @@ type TaskDateTime struct {
...
+	} else {
+		_, err = strconv.ParseUint(s[2:], 10, 32)
+	}
+	if err != nil {
+		return err
+	}
+	*t = AmDuration(s)
+
+	return err
+}
```
Comment (sevein) [inline v2_task.go]: You don't want to use regexp in this code path because of the performance implications. We can get away with some basic byte slice comparisons.

I've written an alternative here including tests: https://go.dev/play/p/Ej1aF6Zv-FC. Perhaps you can include my code in `duration.go` and `duration_test.go`.

File: v2_jobs.go [hunk 5]
```diff
@@ -9,8 +9,11 @@ import (
 
 const jobsBasePath = "api/v2beta/jobs"
 
+// NOTE The jobsDetailsQuery is only available on development branches currently.
+const jobsDetailsQuery = "?detailed"
+
 type JobsService interface {
-	List(context.Context, string, *JobsListRequest) ([]Job, *Response, error)
+	List(context.Context, string, *JobsListRequest, bool) ([]Job, *Response, error)
```
Comment (sevein) [inline v2_jobs.go]: I think that the new option could be part of `JobsListRequest`, e.g.:

```diff
type JobsListRequest struct {
	Microservice string `json:"microservice,omitempty"`
	LinkID       string `json:"link_uuid,omitempty"`
	Name         string `json:"name,omitempty"`
+
+	// Detailed request, encoded as a query parameter.
+	Detailed bool `json:"-"`
}
```

It reduces the number of breaking changes that we're making to the API.

File: v2_jobs.go [hunk 6]
```diff
@@ -79,13 +82,16 @@ func (s *JobStatus) UnmarshalJSON(b []byte) error {
...
+func (s *JobsServiceOp) List(ctx context.Context,
+	ID string,
+	r *JobsListRequest,
+	detail bool) ([]Job, *Response, error) {
+	var path string
+	if detail {
+		path = fmt.Sprintf("%s/%s%s", jobsBasePath, ID, jobsDetailsQuery)
+	} else {
+		path = fmt.Sprintf("%s/%s", jobsBasePath, ID)
+	}
```
Comment (sevein) [inline v2_jobs.go]: If the request struct is modified as I suggested above:

```go
type JobsListRequest struct {
	Microservice string `json:"microservice,omitempty"`
	LinkID       string `json:"link_uuid,omitempty"`
	Name         string `json:"name,omitempty"`

	// Detailed request, encoded as a query parameter.
	Detailed bool `json:"-"`
}
```

Then you should be able to do this:

```go
func (s *JobsServiceOp) List(ctx context.Context, ID string, r *JobsListRequest) ([]Job, *Response, error) {
	path := fmt.Sprintf("%s/%s", jobsBasePath, ID)

	if r.Detailed {
		var err error
		path, err = addOption(path, "detailed", "")
		if err != nil {
			return nil, nil, err
		}
	}
	
	// ... the rest remains unmodified
}
```

And here's how `addOption` should look like:

```go
// addOption adds to the URL string `s` a new query parameter.
// This should be part of the amclient.go file.
func addOption(s, key, value string) (string, error) {
	origURL, err := url.Parse(s)
	if err != nil {
		return "", err
	}

	queryValues := origURL.Query()
	queryValues.Set(key, value)

	origURL.RawQuery = queryValues.Encode()

	return origURL.String(), nil
}
```

File: v2_jobs.go [hunk 7]
```diff
@@ -9,8 +9,11 @@ import (
 
 const jobsBasePath = "api/v2beta/jobs"
 
+// NOTE The jobsDetailsQuery is only available on development branches currently.
+const jobsDetailsQuery = "?detailed"
```
Comment (sevein) [inline v2_jobs.go]: We won't need this.

File: v2_task_test.go [hunk 8]
```diff
@@ -23,21 +23,21 @@ func TestTask_Read(t *testing.T) {
 	"time_created": "2019-06-18T00:00:00",
 	"time_started": "2019-07-18T00:00:00",
 	"time_ended": "2019-08-18T00:00:00",
-	"duration": 4294967295
+	"duration": "4294967295"
```
Comment (sevein) [inline v2_task_test.go]: I don't think Archivematica ever writes the duration in seconds as a string right?

I'm looking at https://github.com/artefactual/archivematica/blob/2c03a79e710d655c3740ddbc4bc62eebd4e4b4fe/src/dashboard/src/components/helpers.py#L163-L170:

It sounds like these are all the potential values we'll see:
1. Signed integer: `1`, `0`, `-1`, ...
2. Empty string: `""`
3. Lower than one: `"< 1"`

**Review by sevein — CHANGES_REQUESTED** [review]

File: v2_task_test.go (hunk 8 above)
Comment (Diogenesoftoronto) [reply v2_task_test.go]: I looked into and it is a bit weird that it can return a negative number as a possibility but nonetheless it does. I do think that we should actually just have a panic or something if we do get a negative number because that seems like something went wrong.

File: v2_task.go (hunk 4 above)
Comment (Diogenesoftoronto) [reply v2_task.go]: I made some slight changes to your suggestion. I really like and appreciate the work that went into these suggestions. I was considering benchmarking the two as well. Will do that later!

File: v2_task_test.go (hunk 8 above)
Comment (Diogenesoftoronto) [reply v2_task_test.go]: what a wacky set of returns

File: utils.go [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package amclient
+
+import "net/url"
+
+func addOpts(s, key, value string) (string, error) {
```
Comment (sevein) [inline utils.go]: This function should be moved to `amclient.go` like I originally suggested. My comment wasn't meant to be arbitrary, `amclient.go` is the container for general-use helpers in this package.

Comment (sevein) [inline utils.go]: My suggestion was to name it `addOption` because it adds one option. `addOpts` is misleading because it looks like it supports adding multiple options, but that's not the case.

File: v2_jobs.go [hunk 10]
```diff
@@ -9,6 +9,8 @@ import (
 
 const jobsBasePath = "api/v2beta/jobs"
 
+const jobsDetailsQuery = "?detailed"
```
Comment (sevein) [inline v2_jobs.go]: Again we won't need this, it can be removed. It's now unused.

File: v2_jobs.go [hunk 11]
```diff
@@ -23,6 +25,9 @@ type JobsListRequest struct {
 	Microservice string `json:"microservice,omitempty"`
 	LinkID       string `json:"link_uuid,omitempty"`
 	Name         string `json:"name,omitempty"`
+
+	// WARNING The Detailed query is only available on development branches currently.
```
Comment (sevein) [inline v2_jobs.go]: I think that precision is important: `Detailed` is not a query, it's a query parameter.

Here's an alternative. We're confident that this will be included in Archivematica 1.16.

```suggestion
	// Detailed instructs the server to return additional data in the response.
	// This feature is available starting from Archivematica 1.16, which was not
	// released at the time this comment was written.
```

File: v2_jobs.go [hunk 12]
```diff
@@ -79,14 +84,20 @@ func (s *JobStatus) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
-type Task struct {
-	ID       string `json:"uuid"`
-	ExitCode uint8  `json:"exit_code"`
-}
+func (s *JobsServiceOp) List(ctx context.Context,
+	ID string,
+	r *JobsListRequest) ([]Job, *Response, error) {
```
Comment (sevein) [inline v2_jobs.go]: This is now a cosmetic change and I must say if we include it it will be the **only** multi-line method declaration in the entire project. Can you revert the change?

```suggestion
func (s *JobsServiceOp) List(ctx context.Context, ID string, r *JobsListRequest) ([]Job, *Response, error) {
```

File: v2_jobs_test.go [hunk 13]
```diff
@@ -36,6 +42,11 @@ func TestJobs_List(t *testing.T) {
 	})
 
 	assert.NilError(t, err)
+	ti, _ := time.Parse("2006-01-02T15:04:05", "2024-01-18T01:27:49")
+
+	var tdt TaskDateTime
+	tdt.Time = ti
```
Comment (sevein) [inline v2_jobs_test.go]: There is an alternative where you don't have to use time.Parse:

```suggestion

	ti := TaskDateTime{
		Time: time.Date(2024, time.January, 18, 1, 27, 49, 0, time.UTC),
	}
	var tdt TaskDateTime
	tdt.Time = ti
```

File: v2_task.go [hunk 14]
```diff
@@ -22,6 +24,56 @@ type TaskDateTime struct {
...
+	ExitCode    uint8        `json:"exit_code"`
+	FileID      string       `json:"file_uuid"`
+	Filename    string       `json:"file_name"`
+	CreatedAt   TaskDateTime `json:"time_created"`
+	StartedAt   TaskDateTime `json:"time_started"`
+	CompletedAt TaskDateTime `json:"time_ended"`
+	Duration    Duration     `json:"duration"`
+}
+
+func (d *Duration) UnmarshalJSON(data []byte) error {
```
Comment (sevein) [inline v2_task.go]: This method should be placed right after the type definition.

File: v2_task.go [hunk 15]
```diff
@@ -22,6 +24,56 @@ type TaskDateTime struct {
...
+	// Check for the special "< 1" case.
+	if bytes.ContainsRune(data, '<') {
```
Comment (sevein) [inline v2_task.go]: I suggested `bytes.Equal(data, []byte("\"< 1\""))` because we're looking for that specific string. In this version you've written, `BROKEN<STRING` would not raise an error, my version however returned:

    strconv.ParseInt: parsing "\"<fwef 1\"": invalid syntax

We should restore this condition:

```suggestion
	if bytes.Equal(data, []byte("\"< 1\"")) {
```

File: v2_task.go [hunk 16]
```diff
@@ -22,6 +24,56 @@ type TaskDateTime struct {
...
+		return nil
+	}
+
+	// Check for the empty string case.
+	if bytes.Equal(data, []byte("\"\"")) {
+		return nil
+	}
+
+	// Parse the integer value.
+	// NOTE Integer values in Python can be larger than 64bit and take up an arbitrary amount of memory but 64bits is of a large enough size that we will not run into it being an issue in 584 million years. Hopefully, no Archivematica process will take that long.
```
Comment (sevein) [inline v2_task.go]: This is very informative but I'd rather see it replaced by a self-documenting test:

```go
		"Invalid integer (too big)": {
			input:     `123456789012345678901234567890123456789`,
			expectErr: true,
		},
```

File: v2_task.go [hunk 17]
```diff
@@ -22,6 +24,56 @@ type TaskDateTime struct {
...
+//
+//  1. Integer values representing the duration in seconds.
+//  2. A specific string literal "< 1", which we interpret as half a second.
+//
+// Examples:
+//   - `{"duration": 60}` unmarshals to 60 seconds,
+//   - `{"duration": "< 1"}` unmarshals to 500 milliseconds.
+//
+// Python function: https://github.com/artefactual/archivematica/blob/2c03a79e710d655c3740ddbc4bc62eebd4e4b4fe/src/dashboard/src/components/helpers.py#L163-L170.
+type Duration time.Duration
```
Comment (sevein) [inline v2_task.go]: I noticed that the tests I wrote in my playground (link: https://go.dev/play/p/Ej1aF6Zv-FC) weren't included. Was there a particular reason for this?

File: v2_jobs_test.go [hunk 18]
```diff
@@ -36,6 +42,11 @@ func TestJobs_List(t *testing.T) {
...
+
```
Comment (sevein) [inline v2_jobs_test.go]: In this test if I understand correctly `JobsListRequest` should have `Detailed` set to `true`, right?

Ideally we would have two subtests, with Detailed set to true and false, confirming that the fields are set to their zero values.

Ideally the test would confirm that the query parameter "detailed" is being properly encoded in the URL.

**Review by sevein — CHANGES_REQUESTED** [review]

File: v2_task.go (hunk 15 above)
Comment (Diogenesoftoronto) [reply v2_task.go]: But it is not possible for the API to return "\"<fwef 1\" as a response. The api can return "< 1", "", or an integer. If the api changes it makes sense to change the client.

File: v2_task.go (hunk 17 above)
Comment (Diogenesoftoronto) [reply v2_task.go]: I didn't think it was in line with the other things we were testing. For example we do not test the other unmarshall functions, I think perhaps adapting the test, such as adding the other options for the api in the current test would fit more with the type of high level tests we have.

File: v2_jobs_test.go (hunk 18 above)
Comment (Diogenesoftoronto) [reply v2_jobs_test.go]: This is a good catch I can add a separate test. That said I would like to add that in a separate PR building on this one. For now I have updated the test to add the detailed field and set it to true.

File: v2_jobs.go (hunk 10 above)
Comment (Diogenesoftoronto) [reply v2_jobs.go]: Hmm weird I am pretty sure I removed this

File: v2_task.go (hunk 16 above)
Comment (Diogenesoftoronto) [reply v2_task.go]: This is great idea I will put it in the next pr with improved tests.

**Review by sevein — APPROVED** [review]
